### PR TITLE
not fully typescript .d.ts

### PR DESCRIPTION
**Pull request by bluelovers**

### CODE REVIEW

**Comment by derduher** [comment]
❤️ I will look at this this weekend. Though I do not have merge permissions I'd be happy to code review. should address #178

File: lib/sitemap.d.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    /**
+     *  Delete url from sitemap
+     *  @param {String} url
+     */
+    del(url: any): number;
+    /**
+     *  Create sitemap xml
+     *  @param {Function}     callback  Callback function with one argument — xml
+     */
+    toXML(callback: any): string;
```
Comment (RyuuGan) [inline lib/sitemap.d.ts]: Using this you are missing the types for the callback, i.e. if your callback will have 2 arguments you can do something like this:

```
export type Callback<E, T> = (err: E, data: T) => void;

toXML(callback: Callback<Error, string>): void;
```


```suggestion
    toXML(callback: Callback<Error, string>): void;
    toXML(): string;
```

File: lib/sitemap.d.ts [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    /**
+     *  Synchronous alias for toXML()
+     *  @return {String}
+     */
+    toString(): string;
+    toGzip(callback?: Function): any;
```
Comment (RyuuGan) [inline lib/sitemap.d.ts]: Actually you can add some typing here also (as in top example)

File: lib/errors.d.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
+/**
+ * URL in SitemapItem does not exists
+ */
+export declare class NoURLError extends Error {
+    constructor(message?: any);
```
Comment (derduher) [inline lib/errors.d.ts]:
```suggestion
    constructor(message?: string);
```

File: lib/errors.d.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+}
+/**
+ * Protocol in URL does not exists
+ */
+export declare class NoURLProtocolError extends Error {
+    constructor(message?: any);
```
Comment (derduher) [inline lib/errors.d.ts]:
```suggestion
    constructor(message?: string);
```

File: lib/errors.d.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+}
+/**
+ * changefreq property in sitemap is invalid
+ */
+export declare class ChangeFreqInvalidError extends Error {
+    constructor(message?: any);
```
Comment (derduher) [inline lib/errors.d.ts]:
```suggestion
    constructor(message?: string);
```

File: lib/errors.d.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+}
+/**
+ * priority property in sitemap is invalid
+ */
+export declare class PriorityInvalidError extends Error {
+    constructor(message?: any);
```
Comment (derduher) [inline lib/errors.d.ts]:
```suggestion
    constructor(message?: string);
```

File: lib/errors.d.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+}
+/**
+ * SitemapIndex target Folder does not exists
+ */
+export declare class UndefinedTargetFolder extends Error {
+    constructor(message?: any);
```
Comment (derduher) [inline lib/errors.d.ts]:
```suggestion
    constructor(message?: string);
```

File: lib/errors.d.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+}
+export declare class InvalidVideoFormat extends Error {
+    constructor(message?: any);
```
Comment (derduher) [inline lib/errors.d.ts]:
```suggestion
    constructor(message?: string);
```

File: lib/errors.d.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+}
+export declare class InvalidVideoDuration extends Error {
+    constructor(message?: any);
```
Comment (derduher) [inline lib/errors.d.ts]:
```suggestion
    constructor(message?: string);
```

File: lib/errors.d.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+}
+export declare class InvalidVideoDescription extends Error {
+    constructor(message?: any);
```
Comment (derduher) [inline lib/errors.d.ts]:
```suggestion
    constructor(message?: string);
```

File: lib/errors.d.ts [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+}
+export declare class InvalidAttrValue extends Error {
+    constructor(key: any, val: any, validator: any);
```
Comment (derduher) [inline lib/errors.d.ts]: key is a string, val is any, validator is a regex. I'm not sure how to specify type as regex in typescript

File: lib/errors.d.ts [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+}
+export declare class InvalidAttr extends Error {
+    constructor(key: any);
+}
+export declare class InvalidNewsFormat extends Error {
+    constructor(message?: any);
```
Comment (derduher) [inline lib/errors.d.ts]:
```suggestion
    constructor(message?: string);
```

File: lib/errors.d.ts [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+}
+export declare class InvalidNewsAccessValue extends Error {
+    constructor(message?: any);
```
Comment (derduher) [inline lib/errors.d.ts]:
```suggestion
    constructor(message?: string);
```

File: lib/sitemap-item.d.ts [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+        root?: builder.XMLElementOrXMLNode;
+        url?: any;
+    });
+    /**
+     *  Create sitemap xml
+     *  @return {String}
+     */
+    toXML(): string;
+    buildVideoElement(video: {
+        thumbnail_loc: any;
```
Comment (derduher) [inline lib/sitemap-item.d.ts]:
```suggestion
        thumbnail_loc: string;
```

File: lib/sitemap-item.d.ts [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+        title: any;
```
Comment (derduher) [inline lib/sitemap-item.d.ts]:
```suggestion
        title: string;
```

File: lib/sitemap-item.d.ts [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+        description: any;
```
Comment (derduher) [inline lib/sitemap-item.d.ts]:
```suggestion
        description: string;
```

File: lib/sitemap-item.d.ts [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+        content_loc?: any;
```
Comment (derduher) [inline lib/sitemap-item.d.ts]:
```suggestion
        content_loc?: string;
```

File: lib/sitemap-item.d.ts [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+        player_loc?: any;
```
Comment (derduher) [inline lib/sitemap-item.d.ts]:
```suggestion
        player_loc?: string;
```

File: lib/sitemap-item.d.ts [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+        duration?: any;
```
Comment (derduher) [inline lib/sitemap-item.d.ts]:
```suggestion
        duration?: string|number;
```

File: lib/sitemap-item.d.ts [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+        expiration_date?: any;
```
Comment (derduher) [inline lib/sitemap-item.d.ts]:
```suggestion
        expiration_date?: string;
```

File: lib/sitemap-item.d.ts [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+        rating?: any;
```
Comment (derduher) [inline lib/sitemap-item.d.ts]:
```suggestion
        rating?: string|number;
```

File: lib/sitemap-item.d.ts [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+        view_count?: any;
```
Comment (derduher) [inline lib/sitemap-item.d.ts]:
```suggestion
        view_count?: string|number;
```

File: lib/sitemap-item.d.ts [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+        publication_date?: any;
```
Comment (derduher) [inline lib/sitemap-item.d.ts]:
```suggestion
        publication_date?: string;
```

File: lib/errors.d.ts (hunk 11 above)
Comment (RyuuGan) [reply lib/errors.d.ts]:
```suggestion
    constructor(key: string, val: any, validator: RegExp);
```

File: lib/sitemap-item.d.ts [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+        family_friendly?: any;
```
Comment (derduher) [inline lib/sitemap-item.d.ts]: I suggest adding a string enum that only allows yes or no
```
declare enum yesno {
  yes = 'yes',
  no = 'no'
}
```
then adding type yesno

File: lib/sitemap-item.d.ts [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+        tag?: string | string[];
+        category?: any;
```
Comment (derduher) [inline lib/sitemap-item.d.ts]:
```suggestion
        category?: string;
```

File: lib/sitemap-item.d.ts [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+        restriction?: any;
```
Comment (derduher) [inline lib/sitemap-item.d.ts]:
```suggestion
        restriction?: string;
        'restriction:relationship'?: string;
```

File: lib/sitemap-item.d.ts [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+        gallery_loc?: any;
+        price?: any;
```
Comment (derduher) [inline lib/sitemap-item.d.ts]:
```suggestion
        price?: string;
        'price:resolution'?: string;
        'price:currency'?: string;
        'price:type'?: string;

```

File: lib/sitemap-item.d.ts [hunk 28]
```diff
@@ -0,0 +1,77 @@
...
+        requires_subscription?: any;
```
Comment (derduher) [inline lib/sitemap-item.d.ts]: see above enum yesno
```suggestion
        requires_subscription?: yesno;
```

File: lib/sitemap-item.d.ts [hunk 29]
```diff
@@ -0,0 +1,77 @@
...
+        uploader?: any;
```
Comment (derduher) [inline lib/sitemap-item.d.ts]:
```suggestion
        uploader?: string;
```

File: lib/sitemap-item.d.ts [hunk 30]
```diff
@@ -0,0 +1,77 @@
...
+        platform?: any;
```
Comment (derduher) [inline lib/sitemap-item.d.ts]:
```
declare enum allowdeny {
  allow = 'allow',
  deny = 'deny'
}
```
```suggestion
        platform?: string;
        'platform:relationship'?: allowdeny;
```

File: lib/sitemap-item.d.ts [hunk 31]
```diff
@@ -0,0 +1,77 @@
...
+        live?: any;
```
Comment (derduher) [inline lib/sitemap-item.d.ts]:
```suggestion
        live?: yesno;
```

File: lib/sitemap-item.d.ts (hunk 24 above)
Comment (RyuuGan) [reply lib/sitemap-item.d.ts]: The type can also just by a strings like 

```
type YES_NO = 'yes' | 'no';
```

File: index.d.ts [hunk 32]
```diff
@@ -0,0 +1,4 @@
+export * from './lib/sitemap';
+import errors = require('./lib/sitemap');
```
Comment (derduher) [inline index.d.ts]: should this be './lib/errors'?

**Comment by bluelovers** [comment]
https://github.com/ekalinin/sitemap.js/pull/183

File: lib/sitemap.d.ts [hunk 33]
```diff
@@ -0,0 +1,143 @@
+/// <reference types="node" />
```
Comment (derduher) [inline lib/sitemap.d.ts]: Ah so that's how you do it.

File: lib/types.d.ts [hunk 34]
```diff
@@ -0,0 +1,91 @@
...
+    license: string;
+    length?: never;
+}
+export interface IVideoItem {
+    thumbnail_loc: string;
+    title: string;
+    description: string;
+    content_loc?: string;
+    player_loc?: string;
+    'player_loc:autoplay': any;
```
Comment (derduher) [inline lib/types.d.ts]:
```suggestion
    'player_loc:autoplay': boolean;
```

File: lib/types.d.ts [hunk 35]
```diff
@@ -0,0 +1,91 @@
...
+export declare const CHANGEFREQ: EnumChangefreq[];
+export declare const enum EnumYesNo {
+    YES = "yes",
+    NO = "no"
+}
+export declare const enum EnumAllowDeny {
+    ALLOW = "allow",
+    DENY = "deny"
+}
+export declare type ICallback<E extends Error, T> = (err: E, data?: T) => void;
```
Comment (derduher) [inline lib/types.d.ts]: function will not always be called with E if working properly. As @RyuuGan pointed out in another PR. Not sure if that means it should be `E|null` or just marked as optional

Comment (bluelovers) [reply lib/types.d.ts]: don't worry ICallback allow input null, this is make sure err is insof Error Class

Comment (derduher) [reply lib/types.d.ts]: 👍

**Comment by derduher** [comment]
@ekalinin depending on whether you are ok with typescript you should merge either this or #183. Skip this if you are ok with typescript.

**Comment by derduher** [comment]
Closing in favor of the now merged full conversion

**Pull request closed without merging**